typed-th-in-ghc-9/article.md

```diff
@@ -0,0 +1,205 @@
+
+Thankfully, the untyped TH interface remains mostly unchanged, and the typed TH interface should only need to use `Code` instead of `TExp` to compile again with some helper functions.
+
+<!-- TODO: Write about motivation as well? -->
+
+## Example GHC 8 code
+
+To make the examples here easier to follow, let us define a small typed Template Haskell module that we wish to work under GHC 8, which we will later port to GHC 9. The code will be quite simple, its purpose is to parse an environment flag into a known data type or stop the compilation altogether if it fails.
+
+We will use the [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so don't forget to load it. You will also need to activate the `TemplateHaskell` language extension.
+
+```hs
+{-# LANGUAGE TemplateHaskell #-}
+
+module TH where
+
+import Control.Monad.IO.Class (liftIO)
+import Language.Haskell.TH.Syntax (Q, TExp)
+import System.Environment (lookupEnv)
+
+data LogLevel
+  = Development  -- ^ Log errors and debug messages
+  | Production  -- ^ Only log errors
+  deriving (Show)
+
+logLevelFromFlag :: Q (TExp LogLevel)
+logLevelFromFlag = do
+  flag <- liftIO $ lookupEnv "LOG_LEVEL"
+  case flag of
+    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+      "DEVELOPMENT" -> [|| Development ||]
+      "PRODUCTION" -> [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+If you try to compile this module with GHC 9, however, this example will fail to compile:
+
+```hs
+>>> :l TH
+
+TH.hs:18:16: error:
+    • Couldn't match type ‘Code m0’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m0 LogLevel
+    • In the Template Haskell quotation [|| Production ||]
+      In the expression: [|| Production ||]
+      In a case alternative: Nothing -> [|| Production ||]
+   |
+18 |     Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
+   |                ^^^^^^^^^^^^^^^^^^
+
+TH.hs:20:24: error:
+    • Couldn't match type ‘Code m1’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m1 LogLevel
+    • In the Template Haskell quotation [|| Development ||]
+      In the expression: [|| Development ||]
+      In a case alternative: "DEVELOPMENT" -> [|| Development ||]
+   |
+20 |       "DEVELOPMENT" -> [|| Development ||]
+   |                        ^^^^^^^^^^^^^^^^^^^
+
+TH.hs:21:23: error:
+    • Couldn't match type ‘Code m2’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m2 LogLevel
+    • In the Template Haskell quotation [|| Production ||]
+      In the expression: [|| Production ||]
+      In a case alternative: "PRODUCTION" -> [|| Production ||]
+   |
+21 |       "PRODUCTION" -> [|| Production ||]
+   |                       ^^^^^^^^^^^^^^^^^^
+Failed, no modules loaded.
```

Comment:
What I do not particularly like about this example is that it results in three errors that are exactly the same. Can’t we come up with something that will just give one error? I think one error is convincing enough and we don’t need to waste space on three errors.

Reply:
I've trimmed the other two errors and added a comment mentioning they are omitted.